### PR TITLE
Assignment9+10

### PR DESCRIPTION
**Pull request by MarkCapka**
@TylerRose8

 Thanks for a great quarter, ya'll.

### CODE REVIEW

**Comment by ZackWard-student** [comment]
Peer reviewing.

**Comment by ZackWard-student** [comment]
1. Implement PingProcess' public Task<PingResult> RunTaskAsync(string hostNameOrAddress) ✔
- First implement public void RunTaskAsync_Success() test method to test PingProcess.RunTaskAsync() using "localhost". ✔
- Do NOT use async/await in this implementation. ✔

2. Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress) ✔

- First implement the public void RunAsync_UsingTaskReturn_Success() test method to test PingProcess.RunAsync() using "localhost" without using async/await. ✔
- Also implement the async public Task RunAsync_UsingTpl_Success() test method to test PingProcess.RunAsync() using "localhost" but this time DO using async/await. ✔

3. Add support for an optional cancellation token to the PingProcess.RunAsync() signature. ✔

- Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() method so an exception is thrown. ✔ 
- Test that, when cancelled from the test method, the exception thrown is an AggregateException ✔
- with a TaskCanceledException inner exception ✔
- using the test methods RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping ✔
- and RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException ✔ respectively

4. Complete/fix AND test async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default) which executes ping for and array of hostNameOrAddresses (which can all be "localhost") in parallel, adding synchronization if needed. ✔ 

- The order of the items in the stdOutput is irrelevent and expected to be intermingled.
- StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔

5. Implement AND test public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token) using Task.Factory.StartNew() and invoking RunProcessInternal with a TaskCreation value of TaskCreationOptions.LongRunning and a TaskScheduler value of TaskScheduler.Current. Returning a Task<PingResult> is also okay. NOTE: This method does NOT use Task.Run.✔

All the Basic stuff looks goo too!

File: Assignment/Assignment/PingProcess.cs [hunk 1]
```diff
@@ -9,63 +9,188 @@
...
+        cancellationToken.ThrowIfCancellationRequested();
+        return taskPing.Result;
+
     }
 
-    async public Task<PingResult> RunAsync(params string[] hostNameOrAddresses)
-    {
-        StringBuilder? stringBuilder = null;
-        ParallelQuery<Task<int>>? all = hostNameOrAddresses.AsParallel().Select(async item =>
+    async public Task<PingResult> RunAsync(System.IProgress<PingResult> progress, CancellationToken cancellationToken = default, params string[] hostNameOrAddresses)
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:136]: You should have more tests around the extra credit implementation to ensure that it is working as expected.

**Comment by mmwoodfo** [comment]
You've got alot of additional files in your git diff that don't need to be there.

**Pull request closed without merging**